website/client/components/challenges/challengeModal.vue

```diff
@@ -233,7 +233,10 @@ export default {
   async mounted () {
     this.groups = await this.$store.dispatch('guilds:getMyGuilds');
     if (this.user.party._id) {
-      let party = await this.$store.dispatch('guilds:getGroup', {groupId: 'party'});
+      let party;
```

Comment:
two things: can you fix the formatting of the `if`? But mainly I'd moev the code to inside the `guilds:getGroup` action so that when it's called the group is fetched from the server only if necessary, so that this code doesn't have to be added everywhere. And add an optional `forceReload` parameter to the action that if true reloads the data from the server even if it's already cached

Reply:
Great idea - though I can't figure out how to return store data from inside an action, because inside actions things like store.state.user.data.party are Observers rather than data that can be returned. I tried reading the Vuex docs but Habitica does things differently (Vuex advises against returning anything from actions at all).

If there's no way to access store data in a returnable format from inside actions, where else could this function go that would make it reusable?

Also what was wrong with the format of the if? I tried to copy the code style I saw elsewhere.

Reply:
what do you need to return?

You can also not return anything and take the updated value using `this.$store...` but in general Observers are fine to return, it just means that Vue is observing them but if you convert them to JSON with JSON.stringify() you'll see that they're actually normal data, so don't worry about them.

Let me know if you have questions and if you get stuck feel free to push the code you have and I'll give a look :)

The if on a single line is fine if it's on its own but when there's also an else it should follow this form:

```
if (condition) {

} else {

}
```

Reply:
Awesome. Let me try to explain what I'm after (trying to put together all of your suggestions) and then this should be good to go.

You suggest quite rightly that we move the if statement to the ```guilds:getGroup``` action so that it doesn't have to be repeated in every component. For that to work though, I think, we'd need to be able to do just this in the component:
```
const party = await this.$store.dispatch('guilds:getGroup', {groupId: 'party'});
```
And the action would then check the store and return that or else return the results of the API call. I tried code like this:
```
export async function getGroup (store, payload) {
  // Unless forceReload is true, if we are looking for the party and it exists in the store, return that
  // Otherwise call the API
  if (payload.groupId === 'party' && store.state.party.data && !payload.forceReload) {
    return store.state.party;
  } else {
    let response = await axios.get(`/api/v3/groups/${payload.groupId}`);
    // @TODO: should we store the active group for modifying?
    // let guilds = response.data.data;
    // store.state.myGuilds = guilds;

    // @TODO: Populate with members, challenges, and invites

    return response.data.data;
  }
}
```
Here's the problem I have right now: if I return ```store.state.party.data```, I can just leave the code in ```challengeModal.vue``` as is. However, this breaks other components (like ```questDetailsModal.vue``` (somehow party.members is no longer an array in the store if I do that).  If I return just ```store.state.party``` then I don't get any errors from other components, but I have to change the lines in challengeModal to this:
```
this.groups.push({
    name: party.data.name,
    _id: party.data._id,
    privacy: 'private',
  });
```
This is the code I've pushed by this point. It seems to work without any errors, but I can't help feeling that it could be done better somehow.

Reply:
Okay I think one problem could be when the party data is not cached you're returning the response but not updating the store, also `store.party` already exists (see client/store/index.js)

```
      party: {
        quest: {},
        members: asyncResourceFactory(),
      },
```

so it looks like you're not returning the correct party data but quest and members, try adding a new `data` object under store.party where you can save the actual party data

Reply:
Ok. While I'm at it, I've noticed that party data is saved in two different overlapping places: ```store.state.party``` and ```store.state.user.data.party```. The latter has an exact copy of the data stored in ```store.state.party.members.data```. That is, a subset of available data about the user's party is stored in two places. If these were consolidated that would make it a lot easier to reason about what is happening here - is there a reason the data is stored twice? Are there cases where this data would be different?

Also, the more I scratch my head with this code, the more I feel like fetching party data should be an action in ```party.js``` rather than ```guilds.js```. That way we wouldn't have to write code to handle the special case of a party in getGroup. Any thoughts?

Reply:
All right, I just pushed another change. I simply take it for granted that the party data will be available in the store, and just omit the option to attach a challenge to a party if it isn't. This means no changes are required to the actions and no additional API calls need to be made. I think this is pretty clean, except that mapping to ```party.members.data``` returns an array, so you'll see at line 301 I had to index it, which is a bit ugly but it didn't look like ```mapState``` could be coerced into returning an index. Let me know what you think!

Reply:
> I simply take it for granted that the party data will be available in the store

What makes it available in the store?

Anyway when you fetch the party again the store should be updated

Reply:
I did an XHR breakpoint on calls to /group/party and the first one is called by ```header/index.vue```, which loads no matter which page you are on, and calls ```party:fetchMembers``` which itself calls the API, and this is how the data ends up in the store (through ```loadAsyncResource```. All further calls are to actions that just return the API response and do not modify the store, so far as I can tell.

This pattern is confusing to me - unless I'm missing something. The pattern followed by party:fetchMembers seems to be the right one - actions should modify the store, and the data should be gotten from the store. But that isn't how the majority of the actions I've been looking at are written.

Reply:
>  The pattern followed by party:fetchMembers seems to be the right one - actions should modify the store, and the data should be gotten from the store.

Yes that's the correct pattern, we're slowly refactoring actions to make them work that way but it'll take some time

> I did an XHR breakpoint on calls to /group/party and the first one is called by header/index.vue, which loads no matter which page you are on, and calls party:fetchMembers which itself calls the API, and this is how the data ends up in the store (through loadAsyncResource)

Okay, but `store.state.party.members` only contains the party members data, not the party data itself. So you should create a new action (yeah, let's put that into actions/party.js) that loads an async resource into `party.data` (or anothe sub property of `store.state.party`) like this

```
export function getParty (store, forceLoad = false) {
  return loadAsyncResource({
    store,
    path: 'party.data',
    url: '/api/v3/groups/party',
    deserialize (response) {
      return response.data.data;
    },
    forceLoad,
  });
}
```

and in `store/index.js`:

```
      party: {
        data: asyncResourceFactory(),
        quest: {},
        members: asyncResourceFactory(),
      },
```

I think you're on the right path but the party data should be under party.data, party.members only include the profile of party members :) If, for any reason, `party.members.data[0]` includes the party data it's a bug and somehow it's leaked from another part of the code but I don't think this is the case or viewing party members would be broken by now

Reply:
Ok this does look like the best plan. One more question then - when should we call ```party:getParty```? In my mind doing it in ```challengeModal``` seems wrong, as the parent page should already have that info. Maybe in ```group.vue```? Or somewhere higher up in the tree so that every component that might need party data can simply access it?

Either way, this code in ```group.vue``` would need to be changed, I think:
```
let group = await this.$store.dispatch('guilds:getGroup', {groupId: this.searchId});

if (this.isParty) {
  this.$store.state.party.data = group;
  this.group = this.$store.state.party.data;
  this.checkForAchievements();
  return;
}

this.$set(this, 'group', group);
```

I'm thinking of something along the following lines:
```
if (this.isParty) {
  const group = await this.$store.dispatch('party:getParty');
  this.group = this.$store.state.party.data;
  this.checkForAchievements();
} else {
  const group = await this.$store.dispatch('guilds:getGroup', {groupId: this.searchId});
  this.$set(this, 'group', group);
}
```

Reply:
Yeah, that sounds good, fetching the party in the challengeModal shouldn't be necessary as it's already loaded in a different place but since we're caching the value calling `await this.$store.dispatch('party:getParty');` once again is not going to cause problems + it could help if one day we decide to stop loading the party on page load. So I'd keep it

Reply:
All right I just pushed these changes. Let me know what you think!

Reply:
Anything look off so far?